tooling/debugger/src/context.rs

```diff
@@ -15,56 +16,138 @@ use nargo::errors::{ExecutionError, Location};
+
+        if parts.is_empty() || parts.len() > 2 {
+            return Err(DebugLocationFromStrError::InvalidDebugLocationString(s.to_string()));
+        }
+
+        fn parse_components(parts: &Vec<&str>) -> Option<DebugLocation> {
+            match parts.len() {
+                1 => {
+                    let opcode_location = OpcodeLocation::from_str(parts[0]).ok()?;
+                    Some(DebugLocation { circuit_id: 0, opcode_location })
+                }
+                2 => {
+                    let circuit_id = parts[0].parse().ok()?;
+                    let opcode_location = OpcodeLocation::from_str(parts[1]).ok()?;
+                    Some(DebugLocation { circuit_id, opcode_location })
+                }
+                _ => unreachable!("`DebugLocation` has too many components"),
+            }
+        }
+
+        parse_components(&parts)
+            .ok_or(DebugLocationFromStrError::InvalidDebugLocationString(s.to_string()))
+    }
+}
+
 #[derive(Debug)]
 pub(super) enum DebugCommandResult {
     Done,
     Ok,
-    BreakpointReached(OpcodeLocation),
+    BreakpointReached(DebugLocation),
     Error(NargoError),
 }
 
+pub struct ExecutionFrame<'a, B: BlackBoxFunctionSolver<FieldElement>> {
+    circuit_id: u32,
+    acvm: ACVM<'a, FieldElement, B>,
+}
+
 pub(super) struct DebugContext<'a, B: BlackBoxFunctionSolver<FieldElement>> {
     acvm: ACVM<'a, FieldElement, B>,
+    current_circuit_id: u32,
     brillig_solver: Option<BrilligSolver<'a, FieldElement, B>>,
+
+    witness_stack: WitnessStack<FieldElement>,
+    acvm_stack: Vec<ExecutionFrame<'a, B>>,
+
+    backend: &'a B,
     foreign_call_executor: Box<dyn DebugForeignCallExecutor + 'a>,
+
     debug_artifact: &'a DebugArtifact,
-    breakpoints: HashSet<OpcodeLocation>,
-    source_to_opcodes: BTreeMap<FileId, Vec<(usize, OpcodeLocation)>>,
+    breakpoints: HashSet<DebugLocation>,
+    source_to_locations: BTreeMap<FileId, Vec<(usize, DebugLocation)>>,
+
+    circuits: &'a [Circuit<FieldElement>],
     unconstrained_functions: &'a [BrilligBytecode<FieldElement>],
 
-    // Absolute (in terms of all the opcodes ACIR+Brillig) addresses of the ACIR
-    // opcodes with one additional entry for to indicate the last valid address.
-    acir_opcode_addresses: Vec<usize>,
+    // We define the address space as a contiguous space where all ACIR and
+    // Brillig opcodes from all circuits are laid out.
+    // The first element of the outer vector will contain the addresses at which
+    // all ACIR opcodes of the first circuit are. The second element will
+    // contain the second circuit and so on. Brillig functions should are
+    // expanded on each ACIR Brillig call opcode.
+    // At the end of each vector (both outer and inner) there will be an extra
+    // sentinel element with the address of the next opcode. This is only to
+    // make bounds checking and working with binary search easier.
+    acir_opcode_addresses: Vec<Vec<usize>>,
```

Comment:
It would be clearer to have this as its own type so that the address of the next opcode is simply a separate field rather than an extra sentinel element in the acir opcode list.

Reply:
The sentinel was initially added (before multiple circuits) to more easily handle the case when the last plus one address was being mapped to the opcode location. Without the sentinel, the function above would return a Brillig opcode location with the Brillig index beyond the unconstrained function's opcodes length. That was incorrect and problematic when handling the returned opcode location.

But we now introduced [a bounds check](https://github.com/manastech/noir/blob/feat%2F4824-multiple-acir-calls/tooling/debugger/src/context.rs#L333), which may as well check against a dedicated sentinel field.